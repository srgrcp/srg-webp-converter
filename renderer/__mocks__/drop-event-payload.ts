export const dropEventPayloadMock = {
  dataTransfer: {
    clearData: () => {},
    items: {
      clear: () => {},
      length: 1,
      0: {
        kind: 'file',
        type: 'image/png',
        getAsFile: () => ({
          path: 'test.png',
          size: 123123,
        }),
      },
    },
  },
}
