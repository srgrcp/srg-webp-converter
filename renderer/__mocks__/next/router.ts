export const push = jest.fn();

export const useRouter = jest.fn().mockImplementation(() => ({
  push,
}));
