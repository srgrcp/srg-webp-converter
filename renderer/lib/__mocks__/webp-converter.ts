import { InputFile } from "@shared/input-file";
import { WebpConverter } from "../webp-converter";

export const webpConverter: WebpConverter = {
  openFiles: () => Promise.resolve<InputFile[]>([
    {
      fileName: "test",
      fileExtension: ".png",
      filePath: "test.png",
      size: '1 MB',
    }
  ]),
  convertFiles: () => Promise.resolve(),
  selectOutputDirectory: () => Promise.resolve('path/to/output/directory'),
}
