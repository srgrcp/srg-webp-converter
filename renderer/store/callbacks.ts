import { useRecoilCallback } from "recoil";
import { webpConverter } from "../lib/webp-converter";
import { converterState } from "./state";

export const useOpenFileDialog = () => {
  return useRecoilCallback(({ set, snapshot }) => async () => {
    const release = snapshot.retain();
    const inputFiles = await webpConverter.openFiles();
    const _converterState = await snapshot.getPromise(converterState);

    set(converterState, ({ ..._converterState, inputFiles }));
    release();
    return inputFiles.length;
  }, []);
};

export const useConvertFiles = () => {
  return useRecoilCallback(({ set, snapshot }) => async () => {
    const release = snapshot.retain();
    const inputFilesState = await snapshot.getPromise(converterState);
    const inputFiles = inputFilesState.inputFiles;
    
    set(converterState, ({ ...inputFilesState, isConverting: true }));
    await webpConverter.convertFiles(inputFiles);
    set(converterState, ({ ...inputFilesState, isConverting: false }));

    release();
  }, []);
}

export const useSelectOutputDirectory = () => {
  return useRecoilCallback(({ set, snapshot }) => async () => {
    const release = snapshot.retain();

    const outputDirectory = await webpConverter.selectOutputDirectory();
    const _converterState = await snapshot.getPromise(converterState);

    set(converterState, ({ ..._converterState, outputDirectory }));
    release();
    return outputDirectory;
  }, []);
}

export const useSetGlobalQuality = () => {
  return useRecoilCallback(({ set, snapshot }) => async (quality: number) => {
    const release = snapshot.retain();
    const _converterState = await snapshot.getPromise(converterState);

    set(converterState, ({ ..._converterState, globalQuality: quality }));
    release();
  }, []);
}

export const useApplyGlobalSettings = () => {
  return useRecoilCallback(({ set, snapshot }) => async () => {
    const release = snapshot.retain();
    const _converterState = await snapshot.getPromise(converterState);

    const inputFiles = _converterState.inputFiles;
    const globalQuality = _converterState.globalQuality;
    const outputDirectory = _converterState.outputDirectory;

    const newInputFiles = inputFiles.map(inputFile => {
      return {
        ...inputFile,
        quality: globalQuality,
        outputDirectory,
      };
    });
    set(converterState, ({ ..._converterState, inputFiles: newInputFiles }));
    release();
  }, []);
}
