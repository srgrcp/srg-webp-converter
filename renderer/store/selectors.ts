import { selector } from "recoil";
import { converterState } from "./state";

export const selectAllInputFiles = selector({
  key: "selectAllInputFiles",
  get: ({ get }) => {
    return get(converterState).inputFiles;
  },
});

export const selectInputFilesLength = selector({
  key: "selectInputFilesLength",
  get: ({ get }) => {
    return get(converterState).inputFiles.length;
  },
});

export const selectOutputDirectory = selector({
  key: "selectOutputDirectory",
  get: ({ get }) => {
    return get(converterState).outputDirectory;
  },
});

export const selectGlobalQuality = selector({
  key: "selectGlobalQuality",
  get: ({ get }) => {
    return get(converterState).globalQuality;
  },
});

export const selectIsConverting = selector({
  key: "selectIsConverting",
  get: ({ get }) => {
    return get(converterState).isConverting;
  },
});
