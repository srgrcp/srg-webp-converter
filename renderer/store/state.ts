import { InputFile } from "@shared/input-file";
import { atom, useRecoilValue } from "recoil";
import { useSetGlobalQuality } from "./callbacks";
import { selectGlobalQuality } from "./selectors";

export type ConverterState = {
  inputFiles: InputFile[];
  outputDirectory: string;
  globalQuality: number;
  isConverting: boolean;
}

export const converterState = atom<ConverterState>({
  key: "converterState",
  default: {
    inputFiles: [],
    outputDirectory: '',
    globalQuality: 80,
    isConverting: false,
  },
});

export const useGlobalQuality = (): [number, (quality: number) => Promise<void>] => {
  return [useRecoilValue(selectGlobalQuality), useSetGlobalQuality()];
}
