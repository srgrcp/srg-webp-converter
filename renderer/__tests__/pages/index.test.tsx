import { fireEvent, render, waitFor } from '@testing-library/react'
import { RecoilRoot } from 'recoil'
import Home from '../../pages'
import { converterState } from '../../store/state'
import { dropEventPayloadMock } from '../../__mocks__/drop-event-payload'
import { push } from '../../__mocks__/next/router'
import { RecoilObserver } from '../../__mocks__/recoil-observer'

jest.mock('../../lib/webp-converter')
jest.mock('next/router')

describe('Home', () => {
  let renderResult: ReturnType<typeof render>
  let dropArea: HTMLElement

  const onStateChange = jest.fn()

  beforeEach(() => {
    renderResult = render(
      <RecoilRoot>
        <RecoilObserver node={converterState} onChange={onStateChange} />
        <Home />
      </RecoilRoot>,
    )
    dropArea = renderResult.getByTestId('drop-files-area')
  })

  describe('snapshot', () => {
    it('should match snapshot', () => {
      const { container } = renderResult
  
      expect(container).toMatchSnapshot()
    })
  })

  describe('user click on drop area to open file dialog', () => {

    it('should update the state with the selected files', async () => {
      // Simulate click on drop area to open file dialog
      fireEvent.click(dropArea)
      
      await waitFor(() => {
        expect(onStateChange).toHaveBeenCalledWith(expect.objectContaining({
          inputFiles: expect.any(Array),
        }))
      })
    })

    it('[click] should redirect to the list page', async () => {
      // Simulate click on drop area to open file dialog
      fireEvent.click(dropArea)

      await waitFor(() => {
        expect(push).toHaveBeenCalledWith('/list')
      })
    })
  })

  describe('user drag and drop files on drop area', () => {

    it('should update the state when files are droped', async () => {
      // Simulate drag and drop files on drop area
      fireEvent.drop(dropArea, dropEventPayloadMock)
      
      await waitFor(() => {
        expect(onStateChange).toHaveBeenCalledWith(expect.objectContaining({
          inputFiles: expect.any(Array),
        }))
      })
    })

    it('[drop] should redirect to the list page', async () => {
      // Simulate drag and drop files on drop area
      fireEvent.drop(dropArea, dropEventPayloadMock)

      await waitFor(() => {
        expect(push).toHaveBeenCalledWith('/list')
      })
    })
  })
})
