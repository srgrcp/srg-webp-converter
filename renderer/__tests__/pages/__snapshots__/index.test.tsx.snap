// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home snapshot should match snapshot 1`] = `
<div>
  <div
    class="container mx-auto py-5 px-4 md:px-8 lg:px-12 2xl:px-16 h-screen flex items-center"
  >
    <div
      class="flex flex-grow flex-col items-center w-full"
    >
      <h1
        class="mb-6 font-bold text-xl text-slate-800 dark:text-slate-300"
      >
        Select images to convert
      </h1>
      <div
        class="w-full max-w-md h-60 border-4 border-dashed border-slate-300 dark:border-slate-900 rounded-xl bg-slate-300/50 dark:bg-slate-900/50 flex flex-col justify-center items-center cursor-pointer"
        data-testid="drop-files-area"
      >
        <p
          class="text-slate-400 font-bold"
        >
          Drag & Drop your files here
        </p>
      </div>
    </div>
  </div>
</div>
`;
